### ISSUE TITLE
Rename "onLockupStreamWithdrawn()" hooks to avoid duplicate calls

### ISSUE THREAD

**Issue by smol-ninja** [sablier-labs/v2-core#822]
### Discussed in https://github.com/sablier-labs/v2-core/discussions/804

<div type='discussions-op-text'>

<sup>Originally posted by **smol-ninja** January 26, 2024</sup>
A note to the reader: **The following doesn't affect the deployed version of Sablier contracts**.

## Context
The recent [PR: Make withdraw function callable by any account](https://github.com/sablier-labs/v2-core/issues/731) enables anyone to call `withdraw()` on any stream. Each withdrawal attempts to call `onLockupStreamWithdrawn()` on both sender and recipient if they are contracts. As discussed privately on Slack, it raises a concern when the stream has the same sender and recipient. [^1]

## Problem
We use the same function name `onLockupStreamWithdrawn()` in both `ISablierV2Recipient` and `ISablierV2Sender`. The user is expected to implement `onLockupStreamWithdrawn()` with no information regarding which interface he wants to implement it for.  This leads to duplicate calls on this function when `withdraw()` is triggered by a third party and the sender and recipient are the same contract addresses.

[SablierV2Lockup.sol#L278C1-L297C10](https://github.com/sablier-labs/v2-core/blob/staging/src/abstracts/SablierV2Lockup.sol#L278C1-L297C10)
```solidity
if (msg.sender != recipient && recipient.code.length > 0) {
    try ISablierV2Recipient(recipient).onLockupStreamWithdrawn({.. /* snip */ ..}) { } catch { }
}

if (msg.sender != sender && sender.code.length > 0) {
    try ISablierV2Sender(sender).onLockupStreamWithdrawn({.. /* snip */ ..}) { } catch { }
}
```

### Option 1
Rename `onLockupStreamWithdrawn()`
- to `callSenderOnLockupStreamWithdrawn()` in `ISablierV2Sender`
- to `callRecipientOnLockupStreamWithdrawn()` in `ISablierV2Recipient` 

Because we have other functions `OnLockupStreamCanceled()` and `onLockupStreamRenounced()` in the `ISablierV2Recipient`, we should also consider renaming them to follow the same naming convention.

### Option 2
When the sender is the same as the recipient, `withdraw()` calls on `onLockupStreamWithdrawn()` only once.

```solidity
if (msg.sender != recipient && recipient.code.length > 0) {
    try ISablierV2Recipient(recipient).onLockupStreamWithdrawn({.. /* snip */ ..}) { } catch { }
}

if (sender != recipient && msg.sender != sender && sender.code.length > 0) {
    try ISablierV2Sender(sender).onLockupStreamWithdrawn({.. /* snip */ ..}) { } catch { }
}
```

[^1]: A user provides liquidity on AMM through a multisig and then locks the liquidity by setting up a stream of LP tokens to the same multisig. The multisig implements `onLockupStreamWithdrawn()` to withdraw liquidity from the AMM every time a `withdraw()` is triggered.</div>

**Comment by smol-ninja** [comment]
I am facing with a few problems while writing test cases for this.

1. when sender == recipient, how to write a test to not expect a call? A test which should detect the above issue.

2. New concrete tests are needed for hook calls. Incorporating them into `withdraw` would be difficult because the basis of the withdraw tree structure is based on whether the withdrawal address is stream recipient or not. In the new test, we can have a structure like this:
    1. When the caller is unknown
        - if sender != recipient 
           - both hook calls should be expected
       - if sender == recipient
           - only one hook call should be expected (how to check this?)
    2. When the caller is the sender
        - if sender != recipient 
           - recipient hook calls should be expected
       - if sender == recipient
           - no hook call should be expected (how to check this?)
     3. When the caller is the recipient
        - if sender != recipient 
           - sender hook calls should be expected
       - if sender == recipient
           - no hook call should be expected (how to check this?)

Any thoughts @andreivladbrg?

**Comment by smol-ninja** [comment]
Thank you for your reply @PaulRBerg.

> You should ask your questions in the Foundry community forums

Totally, I agree. I will ask there.

> By passing the count parameter to vm.expectCall

Oh yes!

**Comment by andreivladbrg** [comment]
> Any thoughts @andreivladbrg?

for 1. PRB already answered, i.e. `count` can be used

regarding 2. i think it would be better to have these branches instead, as it is concise, clearer and shorter:

1. given sender == recipient
   - when caller unknown
      - only one hook call should be expected
   - when the caller is the sender and the recipient
      - no hook call should be expected
2. given sender != recipient
   - when caller unknown
      - both hook calls should be expected
   - when the caller is the sender
      - recipient hook calls should be expected
   - when the caller is the recipient
      - sender hook calls should be expected


> all this is not a matter of internal knowledge at Sablier. You should ask your questions in the Foundry community forums

@PaulRBerg are you sure about this statement? this might only be ok for the question "how to write a test to not expect a call?" but, the rest, what does Foundry have to do with how we write our tree structure for the `withdraw` function?

**Comment by PaulRBerg** [comment]
> @PaulRBerg are you sure about this statement?

Yes. It's always a good idea to talk to the community and correct errors in our understanding. That's the default approach we should take.

https://twitter.com/PaulRBerg/status/1335282362968633344

> what does Foundry have to do with how we write our tree structure for the withdraw function?

I wasn't referrign to that - I was referring to his questions about how to expect calls

**Comment by PaulRBerg** [comment]
My wording was bad, though - I shouldn't have said "all this".

**Comment by smol-ninja** [comment]
> regarding 2. i think it would be better to have these branches instead

I like this more.

> My wording was bad, though - I shouldn't have said "all this". Sorry for that

No problem. I understood your point.
